### ISSUE TITLE
`ValueError` related to option assignment in transactions csv file

### ISSUE THREAD

**Issue by def-mycroft** [beancount/johnny#5]
in the file that I export from tastyworks, I have a couple of lines like this one:

```bash
$  cat tastyworks_transactions_x9678_2016-02-23_2021-06-26.csv | grep "Removal of option due to assignment"
2021-06-25T16:00:00-0500,Receive Deliver,,NKE   210625C00139000,Equity Option,Removal of option due to assignment...
```

...when starting `johnny`, I receive this error message: 

```python
ValueError: Could not infer price from description: ('2021-06-25T16:00:00-0500', KeyError("Invalid rowtype: 'Receive Deliver' (Removal of option due to assignment)"), '', 'NKE   210625C00139000', 'EquityOption', 'Removal of option due to assignment'...
```

**Comment by def-mycroft** [comment]
would be happy to try to apply a fix, thought it best to create a ticket and discuss a bit first though.

**Comment by blais** [comment]
Thanks Joe!
I don't have any assignments in my account history, so I haven't seen this type yet. TW doesn't provide a spec of their file either...

**Comment by Graeme22** [comment]
The way I dealt with this was treating assignment/exercise as a closing trade with a value of zero. They're always followed by an opening trade with the controlled shares.

**Comment by blais** [comment]
Yep, I just want one or two example of a full line with all the data to confirm the numbers.

**Comment by Graeme22** [comment]
Here's a full csv: https://github.com/Graeme22/tastyworks-cli/blob/main/tests/data/transactions.csv

**Comment by def-mycroft** [comment]
cool, looks like the file that @Graeme22 shared [this one](https://github.com/Graeme22/tastyworks-cli/blob/main/tests/data/transactions.csv) has examples that are equivalent to mine.

**Comment by def-mycroft** [comment]
I ended up pulling the price out of the description field too,  in some reporting tools that I built a while ago. it is interesting that we both ended up doing the same thing. leads to issues like this one but are forced to do it this way iirc.

**Comment by blais** [comment]
I'm going to make a detailed report to Tasty about how to improve their transactions log.

@Graeme22 Did you strip the Order # column? Your import fails for  it is missing.
I'll add it back in, but I wonder how come it's not there.

**Comment by blais** [comment]
I'm adding support for crypto with your file @Graeme22 , thanks!

**Comment by blais** [comment]
```
89:2021-02-12T15:52:01-0700,Receive Deliver,BUY_TO_OPEN,ETH/USD,Cryptocurrency,Bought 0.02703492 ETH/USD @ 1849.46 - Awarded 0.02703492 Long ETH/USD,0,0.02703492,0,--,0,,,,,,000000087
```

What does this line correspond to?

**Comment by Graeme22** [comment]
> ```
> 89:2021-02-12T15:52:01-0700,Receive Deliver,BUY_TO_OPEN,ETH/USD,Cryptocurrency,Bought 0.02703492 ETH/USD @ 1849.46 - Awarded 0.02703492 Long ETH/USD,0,0.02703492,0,--,0,,,,,,000000087
> ```
> 
> What does this line correspond to?

A TW promo I believe. They matched crypto up to $50 in value.

**Comment by Graeme22** [comment]
> I'm going to make a detailed report to Tasty about how to improve their transactions log.
> 
> @Graeme22 Did you strip the Order # column? Your import fails for it is missing.
> I'll add it back in, but I wonder how come it's not there.

I did. Pretty sure it's not sensitive but I didn't see a use for it regardless.

**Comment by blais** [comment]
There's a flaw in the way I process opening positions that makes it difficult to test this without a full corresponding positions file.
I realize it's unreasonable or even impossible to provide a matching file for this historical log, so I'm going to review that code.
It's also going to make it more robust.
In the meantime, I'll check in the change.

**Comment by Graeme22** [comment]
Technically the positions files is redundant, as you can calculate the positions from all the unclosed positions in the transaction file. Not sure if that's how you want to go about that though.

**Comment by blais** [comment]
I pushed a bunch of changes; please try again.

I'll go and make change to insert opening balances differently than I currently do (as a diff from the positions file, right now, which it shouldn't require, really).

**Comment by blais** [comment]
@Graeme22 Yes, but something has to be done for positions never opened before.

**Comment by Graeme22** [comment]
> @Graeme22 Yes, but something has to be done for positions never opened before.
> 

Not sure I get what you mean, those will still be in your transactions file no?

**Comment by blais** [comment]
If you had a position before the beginning of the log, and the positions file holds it, its opening has to be synthesized somehow.

**Comment by blais** [comment]
My code so ugly... I have multiple places where I do this inventory calculations.
I'm thinking now of relaxing the requirements on the various transactions logs outputs, and centralizing and inferring all the missing data. There's a lot of crap I'm fixing in normalization... missing expiration, quantities on expirations (and signs and effects), opening missing positions, marking for the close...

**Comment by Graeme22** [comment]
> If you had a position before the beginning of the log, and the positions file holds it, its opening has to be synthesized somehow.
> 
> Basically, if you have only a partial window of time of a transactions log, you have to be ready to reconcile
> 1. positions opened and closed within the window
> 2. positions opened BEFORE the window and closed within the window
> 3. positions opened within the window and never closed (or closed after the window, if the window isn't up to now)
> 4. positions opened before the window, and never closed during the window (or closed after the window)
> 
> (1) is never a problem.
> (2) will require synthesizing a fake opening trade, and can be detected by their unmatched closing
> (3) is going to result in residual positions leftover in the inventory accumulator after processing the log
> (4) is basically undetectable, except if you have a positions file to reconcile against (you will find positions you didn't expect and for those you need to synthesize opens for those)
> 
> 
> 

Ah I see. I solved this by requiring the full transaction history but that becomes more work the longer you trade.

**Comment by blais** [comment]
I solved this by requiring the final positions file and playing the inventory matching backwards, but I now realize it can be made simpler. I'll try to rewrite the whole thing tomorrow (I have a plan, which will make it more robust, also more efficient). I also have to support my TOS account.

Note that this normalization code is something you could reuse in your software BTW. The parsing and normalization code is separate from the presentation layer.

**Comment by pgvee** [comment]
> I'm going to make a detailed report to Tasty about how to improve their transactions log.
> 
> @Graeme22 Did you strip the Order # column? Your import fails for it is missing.
> I'll add it back in, but I wonder how come it's not there.

Hi @blais I'm also working on importing my TW transactions and noticed the order# column is missing on assignment. here's the csv i'm working with (feel free to download it): https://pastebin.com/fXJjNAy0 (the paste will evaporate in a few days).

**Comment by blais** [comment]
@pgvee I just make up an order id.
https://github.com/beancount/johnny/blob/master/johnny/sources/tastyworks_api/transactions.py#L212
